File: src/transfer/transfer_sender.cpp
#include "transfer/transfer_sender.h"
#include "common/protocol.h"
#include "common/utils.h"
#include "common/peer_key.h"
#include <iostream>
#include <fstream>
#include <chrono>
#include <algorithm>
#include <cstring>
#include <netinet/tcp.h>
#include <unistd.h>
#include <arpa/inet.h>

TransferSender::TransferSender()
    : m_transferring(false)
{
}

TransferSender::~TransferSender() {
}

bool TransferSender::send_file(const std::string& target_ip,
                                uint16_t target_port,
                                const std::string& file_path,
                                const std::string& file_id,
                                uint32_t num_threads,
                                ProgressCallback callback) {
    m_transferring = true;
    m_target_ip = target_ip;
    m_metrics = TransferMetrics{};

    FileChunkIO io(file_path, Defaults::CHUNK_SIZE, true);
    if (io.get_file_size() == 0) {
        std::cerr << "[发送] 文件为空或无法打开: " << file_path << std::endl;
        m_transferring = false;
        return false;
    }

    FileMeta meta;
    meta.file_id      = file_id;
    meta.filename     = Utils::get_filename(file_path);
    meta.chunk_size   = Defaults::CHUNK_SIZE;

    std::string temp_path;
    bool compressed = false;
    {
        std::ifstream src(file_path, std::ios::binary);
        std::string raw((std::istreambuf_iterator<char>(src)),
                        std::istreambuf_iterator<char>());
        src.close();

        std::string compressed_data = Utils::compress_data(raw);
        if (!compressed_data.empty() && compressed_data.size() < raw.size()) {
            temp_path = "/tmp/p2p_send_compressed_" + file_id;
            std::ofstream tmp(temp_path, std::ios::binary);
            tmp.write(compressed_data.data(), compressed_data.size());
            tmp.close();
            meta.file_size    = compressed_data.size();
            meta.total_chunks = (static_cast<uint32_t>(compressed_data.size()) + Defaults::CHUNK_SIZE - 1)
                                / Defaults::CHUNK_SIZE;
            compressed = true;
            std::cout << "[发送] zlib 压缩: " << Utils::format_file_size(raw.size())
                      << " → " << Utils::format_file_size(meta.file_size)
                      << " (" << (100 - meta.file_size * 100 / raw.size()) << "% 节省)" << std::endl;
        } else {
            meta.file_size    = io.get_file_size();
            meta.total_chunks = io.get_total_chunks();
        }
    }

    std::string send_path = compressed ? temp_path : file_path;

    if (meta.total_chunks < num_threads) {
        num_threads = meta.total_chunks > 0 ? meta.total_chunks : 1;
    }

    std::cout << "[发送] 准备发送: " << meta.filename
              << " (" << Utils::format_file_size(meta.file_size) << ")"
              << ", 分片: " << meta.total_chunks
              << ", " << num_threads << " 线程" << std::endl;

    auto start_time = std::chrono::steady_clock::now();

    std::vector<std::thread> workers;
    std::atomic<uint64_t> global_bytes_sent(0);
    std::atomic<uint32_t> global_chunks_sent(0);
    std::vector<ThreadMetrics> thread_metrics(num_threads);

    for (uint32_t t = 0; t < num_threads; ++t) {
        uint32_t chunks_per = meta.total_chunks / num_threads;
        uint32_t rem = meta.total_chunks % num_threads;
        uint32_t start = t * chunks_per + std::min(t, rem);
        uint32_t end = start + chunks_per + (t < rem ? 1 : 0);

        workers.emplace_back([this, target_ip, target_port, send_path, meta,
                              start, end, t, num_threads,
                              &global_bytes_sent, &global_chunks_sent, callback,
                              &thread_metrics, start_time]() {
            send_chunk_range(target_ip, target_port, send_path, meta,
                             start, end, t, num_threads,
                             global_bytes_sent, global_chunks_sent,
                             thread_metrics[t]);
        });
    }

    if (callback) {
        while (global_chunks_sent.load() < meta.total_chunks) {
            std::this_thread::sleep_for(std::chrono::milliseconds(200));
            auto now = std::chrono::steady_clock::now();
            auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time).count();
            uint64_t bs = global_bytes_sent.load();
            double speed = elapsed > 0 ? (bs * 1000.0 / elapsed) : 0;
            TransferProgress p;
            p.total_chunks = meta.total_chunks;
            p.sent_chunks  = global_chunks_sent.load();
            p.acked_chunks = global_chunks_sent.load();
            p.bytes_sent   = bs;
            p.speed        = speed;
            p.elapsed_ms   = elapsed;
            callback(p);
        }
    }

    for (auto& w : workers) w.join();

    auto end_time = std::chrono::steady_clock::now();
    auto total_us = std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time).count();

    if (compressed && !temp_path.empty()) {
        std::remove(temp_path.c_str());
    }

    m_metrics.file_size      = meta.file_size;
    m_metrics.num_threads    = num_threads;
    m_metrics.total_time_us  = static_cast<uint64_t>(total_us);
    m_metrics.throughput_mbps = total_us > 0 ? (meta.file_size * 8.0 / total_us) : 0;
    m_metrics.threads         = std::move(thread_metrics);

    m_transferring = false;

    std::cout << "[发送] 传输完成, 耗时 " << (total_us / 1000) << " ms, "
              << m_metrics.throughput_mbps << " Mbps, "
              << num_threads << " 线程" << std::endl;

    {
        json j;
        j["file_size"] = m_metrics.file_size;
        j["num_threads"] = m_metrics.num_threads;
        j["total_time_us"] = m_metrics.total_time_us;
        j["throughput_mbps"] = m_metrics.throughput_mbps;
        json threads_arr = json::array();
        for (auto& tm : m_metrics.threads) {
            json t;
            t["connection_time_us"] = tm.connection_time_us;
            t["first_byte_time_us"] = tm.first_byte_time_us;
            t["total_time_us"] = tm.total_time_us;
            t["bytes_sent"] = tm.bytes_sent;
            t["start_chunk"] = tm.start_chunk;
            t["end_chunk"] = tm.end_chunk;
            json samples = json::array();
            for (auto& s : tm.speed_samples) {
                samples.push_back({s.first, s.second});
            }
            t["speed_samples"] = samples;
            threads_arr.push_back(t);
        }
        j["threads"] = threads_arr;
        std::ofstream metrics_file("/tmp/p2p_metrics.json");
        metrics_file << j.dump(2);
        metrics_file.close();
    }

    return true;
}

bool TransferSender::send_file_header(SOCKET_FD sock, const FileMeta& meta) {
    json header;
    header["type"]         = "FILE_HEADER";
    header["file_id"]      = meta.file_id;
    header["filename"]     = meta.filename;
    header["file_size"]    = meta.file_size;
    header["chunk_size"]   = meta.chunk_size;
    header["total_chunks"] = meta.total_chunks;
    header["compression"]  = "zlib";

    std::string secret = PeerKey::get(m_target_ip);
    if (!secret.empty()) {
        return Protocol::encrypted_send_json(sock, header, m_target_ip);
    }

    char type_marker = 'J';
    if (SOCK_SEND(sock, &type_marker, 1, 0) != 1) return false;
    return Protocol::send_json_message(sock, header);
}

static bool send_range_header(SOCKET_FD sock, const FileMeta& meta,
                               uint32_t start_chunk, uint32_t end_chunk,
                               uint32_t conn_index, uint32_t total_connections,
                               const std::string& target_ip) {
    json hdr;
    hdr["type"]              = "FILE_RANGE";
    hdr["file_id"]           = meta.file_id;
    hdr["filename"]          = meta.filename;
    hdr["file_size"]         = meta.file_size;
    hdr["chunk_size"]        = meta.chunk_size;
    hdr["total_chunks"]      = meta.total_chunks;
    hdr["compression"]       = "zlib";
    hdr["start_chunk"]       = start_chunk;
    hdr["end_chunk"]         = end_chunk;
    hdr["conn_index"]        = conn_index;
    hdr["total_connections"] = total_connections;

    std::string secret = PeerKey::get(target_ip);
    if (!secret.empty()) {
        char type_marker = 'E';
        if (SOCK_SEND(sock, &type_marker, 1, 0) != 1) return false;
        return Protocol::encrypted_send_json(sock, hdr, target_ip);
    }
    char type_marker = 'J';
    if (SOCK_SEND(sock, &type_marker, 1, 0) != 1) return false;
    return Protocol::send_json_message(sock, hdr);
}

bool TransferSender::send_chunk(SOCKET_FD sock, const Chunk& chunk) {
    std::vector<uint8_t> data = chunk.serialize();

    std::string secret = PeerKey::get(m_target_ip);
    if (!secret.empty()) {
        char type_marker = 'D';
        if (SOCK_SEND(sock, &type_marker, 1, 0) != 1) return false;
        return Protocol::encrypted_send_data(sock, data.data(), data.size(), m_target_ip);
    }

    char type_marker = 'C';
    if (SOCK_SEND(sock, &type_marker, 1, 0) != 1) return false;

    size_t total_sent = 0;
    while (total_sent < data.size()) {
        auto n = SOCK_SEND(sock,
            reinterpret_cast<const char*>(data.data()) + total_sent,
            static_cast<int>(data.size() - total_sent), 0);
        if (n <= 0) return false;
        total_sent += n;
    }
    return true;
}

bool TransferSender::send_chunk_range(const std::string& target_ip,
                                       uint16_t target_port,
                                       const std::string& file_path,
                                       const FileMeta& meta,
                                       uint32_t start_chunk,
                                       uint32_t end_chunk,
                                       uint32_t conn_index,
                                       uint32_t total_connections,
                                       std::atomic<uint64_t>& bytes_sent,
                                       std::atomic<uint32_t>& chunks_sent,
                                       ThreadMetrics& metrics) {
    metrics = ThreadMetrics{};
    metrics.start_chunk = start_chunk;
    metrics.end_chunk   = end_chunk;

    auto t0 = std::chrono::steady_clock::now();

    SOCKET_FD sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET_FD) {
        std::cerr << "[发送线程" << conn_index << "] socket() 失败" << std::endl;
        return false;
    }

    int send_buf = 256 * 1024;
    setsockopt(sock, SOL_SOCKET, SO_SNDBUF, (const char*)&send_buf, sizeof(send_buf));

    struct timeval tv;
    tv.tv_sec = 1;
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&tv, sizeof(tv));

    if (NetworkUtils::enable_bbr(sock)) {
        std::cout << "[发送线程" << conn_index << "] BBR 已启用" << std::endl;
    }

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(target_port);
    inet_pton(AF_INET, target_ip.c_str(), &addr.sin_addr);

    if (connect(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        std::cerr << "[发送线程" << conn_index << "] connect() 失败: "
                  << NetworkUtils::get_last_error_string() << std::endl;
        CLOSE_SOCKET(sock);
        return false;
    }

    auto t_connected = std::chrono::steady_clock::now();
    metrics.connection_time_us = std::chrono::duration_cast<std::chrono::microseconds>(t_connected - t0).count();

    if (!send_range_header(sock, meta, start_chunk, end_chunk, conn_index, total_connections, m_target_ip)) {
        std::cerr << "[发送线程" << conn_index << "] 发送范围头失败" << std::endl;
        CLOSE_SOCKET(sock);
        return false;
    }

    FileChunkIO io(file_path, Defaults::CHUNK_SIZE, true);
    uint64_t local_bytes = 0;
    auto first_sent = false;
    auto last_sample = t_connected;

    for (uint32_t i = start_chunk; i < end_chunk; ++i) {
        Chunk chunk;
        if (!io.read_chunk(i, meta.file_id, meta.total_chunks, chunk)) {
            std::cerr << "[发送线程" << conn_index << "] read_chunk 失败: " << i << std::endl;
            CLOSE_SOCKET(sock);
            return false;
        }
        if (!send_chunk(sock, chunk)) {
            std::cerr << "[发送线程" << conn_index << "] send_chunk 失败: " << i << std::endl;
            CLOSE_SOCKET(sock);
            return false;
        }

        local_bytes += chunk.data_size;
        bytes_sent.fetch_add(chunk.data_size, std::memory_order_relaxed);
        chunks_sent.fetch_add(1, std::memory_order_relaxed);

        if (!first_sent) {
            first_sent = true;
            auto t_first = std::chrono::steady_clock::now();
            metrics.first_byte_time_us = std::chrono::duration_cast<std::chrono::microseconds>(t_first - t0).count();
        }

        auto now = std::chrono::steady_clock::now();
        auto since = std::chrono::duration_cast<std::chrono::microseconds>(now - last_sample).count();
        if (since >= 100000) {
            auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(now - t_connected).count();
            double speed = elapsed > 0 ? (local_bytes * 1000000.0 / elapsed) : 0;
            metrics.speed_samples.push_back({static_cast<uint64_t>(elapsed), speed});
            last_sample = now;
        }
    }

    json done_msg;
    done_msg["type"]     = "RANGE_DONE";
    done_msg["file_id"]  = meta.file_id;
    done_msg["conn_index"] = conn_index;
    done_msg["start_chunk"] = start_chunk;
    done_msg["end_chunk"]   = end_chunk;

    std::string secret = PeerKey::get(m_target_ip);
    if (!secret.empty()) {
        char type_marker = 'E';
        SOCK_SEND(sock, &type_marker, 1, 0);
        Protocol::encrypted_send_json(sock, done_msg, m_target_ip);
    } else {
        char type_marker = 'J';
        SOCK_SEND(sock, &type_marker, 1, 0);
        Protocol::send_json_message(sock, done_msg);
    }

    {
        json ack;
        if (!secret.empty()) {
            char marker;
            if (SOCK_RECV(sock, &marker, 1, 0) == 1 && (marker == 'E' || marker == 'J')) {
                if (marker == 'E') {
                    Protocol::encrypted_recv_json(sock, ack, m_target_ip);
                } else {
                    Protocol::recv_json_message(sock, ack);
                }
            }
        } else {
            char marker;
            if (SOCK_RECV(sock, &marker, 1, 0) == 1 && marker == 'J') {
                Protocol::recv_json_message(sock, ack);
            }
        }
    }

    auto t_end = std::chrono::steady_clock::now();
    metrics.total_time_us  = std::chrono::duration_cast<std::chrono::microseconds>(t_end - t0).count();
    metrics.bytes_sent     = local_bytes;

    CLOSE_SOCKET(sock);
    return true;
}

File: python/p2p_transfer/transfer/transfer_sender.py
"""
文件传输发送端模块

实现多线程并发文件发送，核心特性：

1. 分块范围划分：将 total_chunks 均匀分配到 N 个线程
2. 每个线程独立打开 TCP 连接，互不干扰
3. 可选 zlib 压缩：整个文件压缩后若更小则使用压缩版本
4. 启用 BBR 拥塞控制（Linux 内核 4.9+，静默回退）
5. 传输指标收集（连接耗时、首字节时间、速度采样）

与 C++ 实现的差异：
- 使用 Python threading 替代 C++ std::thread
- 不会为空白范围（empty range）创建线程，避免超发 total_connections
- 指标输出到 /tmp/p2p_metrics.json
"""

import socket
import threading
import time
import os
import math
import json
from p2p_transfer.common.types import (
    TRANSFER_PORT,
    CHUNK_SIZE,
    NUM_THREADS,
    MsgType,
    TransferState,
    FileMeta,
)
from p2p_transfer.common.platform import enable_bbr
from p2p_transfer.common.peer_key import PeerKey
from p2p_transfer.common.protocol import (
    send_json_with_marker,
    recv_json_any,
    send_chunk_data,
    build_file_range,
    build_range_done,
)
from p2p_transfer.common.utils import compress_data, get_filename
from p2p_transfer.transfer.chunk import serialize_chunk
from p2p_transfer.transfer.file_io import FileChunkIO


class ThreadMetrics:
    """单个发送线程的性能指标收集器

    记录连接时间、首字节时间、发送字节数/块数、速度采样点。
    """

    def __init__(self):
        self.connection_time_us = 0         # TCP 连接耗时（微秒）
        self.first_byte_time_us = 0         # 首字节发送时间（微秒自连接开始）
        self.total_time_us = 0              # 总耗时（微秒）
        self.bytes_sent = 0                 # 已发送字节数
        self.chunks_sent = 0                # 已发送分块数
        self.speed_samples: list[tuple[float, float]] = []  # (时间戳, 累积字节)


class TransferSender:
    """多线程文件传输发送器

    使用方法：
        sender = TransferSender()
        sender.send_file(target_ip, port, file_path, file_id, num_threads=4)
    """

    def __init__(self):
        self._num_threads = NUM_THREADS

    def send_file(
        self,
        target_ip: str,
        target_port: int,
        file_path: str,
        file_id: str,
        num_threads: int = NUM_THREADS,
        progress_callback=None,
    ) -> bool:
        """向目标设备发送文件

        流程：
        1. 读取源文件并尝试 zlib 压缩
        2. 若压缩后更小则使用压缩文件，否则使用原文件
        3. 计算分块数，按线程数分配范围
        4. 过滤空白范围，计算实际连接数
        5. 为每个有效范围创建一个发送线程
        6. 等待所有线程完成
        7. 写出指标文件并清理临时压缩文件

        Args:
            target_ip: 目标设备 IP
            target_port: 目标传输端口（通常 8890）
            file_path: 源文件绝对路径
            file_id: 传输任务的 UUID
            num_threads: 并行线程数
            progress_callback: 进度回调 callback(chunks_done, bytes_sent)

        Returns:
            所有线程是否成功完成
        """
        self._num_threads = num_threads

        file_name = get_filename(file_path)
        original_size = os.path.getsize(file_path)
        total_chunks = max(1, math.ceil(original_size / CHUNK_SIZE))
        compressed = False
        actual_path = file_path
        actual_size = original_size

        try:
            # ── 压缩决策 ──
            with open(file_path, "rb") as f:
                file_data = f.read()
            compressed_data = compress_data(file_data)
            if len(compressed_data) < len(file_data):
                # 压缩后更小，写入临时压缩文件并使用
                compressed = True
                compressed_path = f"/tmp/p2p_send_compressed_{file_id}"
                actual_path = compressed_path
                actual_size = len(compressed_data)
                total_chunks = max(1, math.ceil(actual_size / CHUNK_SIZE))
                with open(compressed_path, "wb") as f:
                    f.write(compressed_data)

            # ── 范围划分与线程创建 ──
            # 计算原始分配范围
            chunk_ranges = self._divide_chunks(total_chunks, self._num_threads)
            # 过滤掉 end < start 的空白范围，避免发送无效连接
            active_ranges = [(i, s, e) for i, (s, e) in enumerate(chunk_ranges) if s <= e]
            actual_connections = len(active_ranges)
            if actual_connections == 0:
                return False

            threads = []
            metrics_list = [ThreadMetrics() for _ in range(actual_connections)]
            errors = []                               # 线程错误收集

            for conn_i, (orig_i, start, end) in enumerate(active_ranges):
                t = threading.Thread(
                    target=self._send_chunk_range,
                    args=(
                        target_ip, target_port, actual_path, file_id,
                        file_name, actual_size, total_chunks,
                        start, end, conn_i, actual_connections,
                        compressed, metrics_list[conn_i], errors,
                    ),
                    daemon=True,
                )
                t.start()
                threads.append(t)

            # ── 进度报告（如果提供了回调） ──
            if progress_callback:
                while any(t.is_alive() for t in threads):
                    total_bytes = sum(m.bytes_sent for m in metrics_list)
                    total_chunks_done = sum(m.chunks_sent for m in metrics_list)
                    progress_callback(total_chunks_done, total_bytes)
                    time.sleep(0.2)

            # 等待所有线程完成
            for t in threads:
                t.join(timeout=30)

            # 写出指标文件
            self._write_metrics(metrics_list, target_ip, file_name, actual_size)

            # 清理临时压缩文件
            if compressed and os.path.exists(actual_path) and actual_path != file_path:
                try:
                    os.unlink(actual_path)
                except:
                    pass

            return len(errors) == 0

        except Exception as e:
            print(f"[TransferSender] Error: {e}")
            return False

    def _divide_chunks(self, total_chunks: int, n_threads: int) -> list[tuple[int, int]]:
        """将总分块数均匀分配到 n_threads 个线程

        使用公平分配算法：
        - 基础分配 = total_chunks // n_threads
        - 余数分配给前 rem 个线程
        - 若某个线程分配数为 0，设范围为 (0, -1) 表示空

        Returns:
            列表，每个元素为 (start_chunk, end_chunk)
        """
        ranges = []
        base = total_chunks // n_threads          # 每个线程基础块数
        rem = total_chunks % n_threads             # 剩余块数（分配给前几个线程）
        start = 0
        for i in range(n_threads):
            extra = 1 if i < rem else 0
            count = base + extra
            if count == 0:
                ranges.append((0, -1))             # 空范围：标记为无效
            else:
                end = start + count - 1
                ranges.append((start, end))
                start = end + 1
        return ranges

    def _send_chunk_range(
        self,
        target_ip: str,
        target_port: int,
        file_path: str,
        file_id: str,
        file_name: str,
        file_size: int,
        total_chunks: int,
        start_chunk: int,
        end_chunk: int,
        conn_index: int,
        total_connections: int,
        compressed: bool,
        metrics: ThreadMetrics,
        errors: list,
    ):
        """单个线程的发送逻辑

        独立 TCP 连接，发送指定范围的分块，
        完成后等待 RANGE_ACK 确认。

        有线协议顺序：
        ┌─────────────┬──────────────┬────────────┬────────────┬──────────┐
        │ FILE_RANGE  │ chunk_0..N-1 │ RANGE_DONE │ wait       │ close    │
        │ (J/E JSON)  │ (C/D binary) │ (J/E JSON) │ RANGE_ACK  │          │
        └─────────────┴──────────────┴────────────┴────────────┴──────────┘
        """
        # 防御性检查：空范围跳过
        if start_chunk > end_chunk or end_chunk < 0:
            return

        sock = None
        connect_start = time.time()
        try:
            # ── 连接与配置 ──
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 256 * 1024)  # 256KB 发送缓冲
            sock.settimeout(10)

            sock.connect((target_ip, target_port))
            metrics.connection_time_us = int((time.time() - connect_start) * 1_000_000)

            # 尝试启用 BBR 拥塞控制
            enable_bbr(sock)

            # ── 发送 FILE_RANGE 头部 ──
            comp_str = "zlib" if compressed else ""
            header = build_file_range(
                file_id, file_name, file_size, total_chunks, CHUNK_SIZE,
                start_chunk, end_chunk, conn_index, total_connections, comp_str,
            )
            send_json_with_marker(sock, header, target_ip)

            # ── 打开文件并发送分块 ──
            file_io = FileChunkIO(file_path, is_sender=True, file_id=file_id)

            sock.settimeout(5)
            first_chunk = True

            for chunk_idx in range(start_chunk, end_chunk + 1):
                data = file_io.read_chunk(chunk_idx)
                if data is None:
                    break
                # 序列化为有线格式
                chunk_bytes = serialize_chunk(chunk_idx, total_chunks, file_id, data)
                send_chunk_data(sock, chunk_bytes, target_ip)        # 自动加密

                # 记录首字节时间
                if first_chunk:
                    metrics.first_byte_time_us = int(
                        (time.time() - connect_start) * 1_000_000
                    )
                    first_chunk = False

                metrics.bytes_sent += len(data)
                metrics.chunks_sent += 1

                # 速度采样（每 100ms 一个采样点）
                now = time.time()
                if metrics.speed_samples:
                    last_t, last_b = metrics.speed_samples[-1]
                    if now - last_t >= 0.1:
                        metrics.speed_samples.append((now, metrics.bytes_sent))
                else:
                    metrics.speed_samples.append((now, 0))

            file_io.close()

            # ── 发送 RANGE_DONE 并等待 RANGE_ACK ──
            done_msg = build_range_done(file_id, conn_index)
            send_json_with_marker(sock, done_msg, target_ip)

            sock.settimeout(10)
            ack = recv_json_any(sock, target_ip)
            if ack is None or ack.get("type") != MsgType.RANGE_ACK:
                raise Exception(f"No valid RANGE_ACK received, got: {ack}")

        except Exception as e:
            errors.append(str(e))
        finally:
            if sock:
                try:
                    sock.close()
                except:
                    pass

    def _write_metrics(
        self, metrics_list: list[ThreadMetrics], target_ip: str, file_name: str, file_size: int
    ):
        """将各线程的传输指标汇总写入 /tmp/p2p_metrics.json"""
        summary = {
            "target": target_ip,
            "file": file_name,
            "size": file_size,
            "threads": [],
        }
        total_bytes = 0
        for mi, m in enumerate(metrics_list):
            tinfo = {
                "index": mi,
                "bytes": m.bytes_sent,
                "chunks": m.chunks_sent,
                "connect_us": m.connection_time_us,     # 连接耗时（微秒）
                "first_byte_us": m.first_byte_time_us,   # 首字节时间（微秒）
            }
            total_bytes += m.bytes_sent
            summary["threads"].append(tinfo)

        summary["total_bytes"] = total_bytes
        try:
            with open("/tmp/p2p_metrics.json", "w") as f:
                json.dump(summary, f, indent=2)
        except:
            pass
